host/class/uac/usb_host_uac/uac_host.c

```diff
@@ -2252,22 +2496,103 @@ esp_err_t uac_host_device_set_volume(uac_host_device_handle_t uac_dev_handle, ui
+    // Round to the nearest float value based the vol_res_db
+    volume_db_f = round(volume_db_f / _volume_db_u16_2_f(iface->vol_res_db)) * _volume_db_u16_2_f(iface->vol_res_db);
+    // Convert back to 16-bit value
+    uint16_t volume_db = _volume_db_f_2_u16(volume_db_f);
+    // 0x8000 is a special value means infinity, witch is CUR attribute only
+    volume_db = (volume_db == 0x8000) ? 0x8001 : volume_db;
+
+    UAC_GOTO_ON_ERROR(uac_cs_request_set_volume(iface, volume_db), "Unable to set volume");
+    // Backup the volume value for the get volume function
+    iface->cur_vol = volume;
+    ESP_LOGI(TAG, "Set volume %d%%, Interface %d-%d", volume, iface->dev_info.iface_num, iface->cur_alt + 1);
+    uac_host_interface_unlock(iface);
+    return ESP_OK;
 
-    uint32_t volume_db = volume * UAC_SPK_VOLUME_STEP + UAC_SPK_VOLUME_MIN;
-    UAC_RETURN_ON_ERROR(uac_cs_request_set_volume(iface, volume_db), "Unable to set volume");
+fail:
+    uac_host_interface_unlock(iface);
+    return ret;
+}
+
+esp_err_t uac_host_device_get_volume(uac_host_device_handle_t uac_dev_handle, uint8_t *volume)
+{
+    uac_iface_t *iface = get_iface_by_handle(uac_dev_handle);
+    UAC_RETURN_ON_INVALID_ARG(iface);
+    UAC_RETURN_ON_INVALID_ARG(volume);
+    // Check if the device is active or ready
+    esp_err_t ret = ESP_OK;
+    UAC_RETURN_ON_ERROR(uac_host_interface_try_lock(iface, DEFAULT_CTRL_XFER_TIMEOUT_MS), "Unable to lock UAC Interface");
+    UAC_GOTO_ON_FALSE((UAC_INTERFACE_STATE_ACTIVE == iface->state || UAC_INTERFACE_STATE_READY == iface->state),
+                      ESP_ERR_INVALID_STATE, "device not ready or active");
+
+    // Return the backup volume value to avoid the volume reads differently than expected.
+    // Because the device volume adjustment step may relatively large,
+    // For example, when the user sets 83%, the actual volume may be 80.
+    if (iface->cur_vol) {
+        *volume = iface->cur_vol;
+        uac_host_interface_unlock(iface);
+        return ESP_OK;
+    }
+
+    // Otherwise, get the volume from the device
+    // Get volume range, calculate in dB float
+    uint16_t volume_db = 0;
+    UAC_GOTO_ON_ERROR(uac_cs_request_get_volume(iface, &volume_db), "Unable to get volume");
+    float volume_db_f = _volume_db_u16_2_f(volume_db);
+    // Calculate volume in percentage
+    *volume = (uint8_t)round((volume_db_f - _volume_db_u16_2_f(iface->vol_min_db)) * 100.0 / (_volume_db_u16_2_f(iface->vol_max_db) - _volume_db_u16_2_f(iface->vol_min_db)));
+
+    uac_host_interface_unlock(iface);
     return ESP_OK;
+
+fail:
+    uac_host_interface_unlock(iface);
+    return ret;
 }
 
 esp_err_t uac_host_device_set_volume_db(uac_host_device_handle_t uac_dev_handle, uint32_t volume_db)
 {
     uac_iface_t *iface = get_iface_by_handle(uac_dev_handle);
     UAC_RETURN_ON_INVALID_ARG(iface);
```

Comment:
Should we also check for `volume_db` range here? Similarly to: 
`UAC_RETURN_ON_FALSE(volume <= 100, ESP_ERR_INVALID_ARG, "Invalid volume value");`
in `uac_host_device_set_volume()`

Reply:
Yes, you are right, let's check it below